https://assaeunji.github.io/machine%20learning/2020-11-29-implicitfeedback/

# test 3주 정합성 검증
- 

# 추천 모델
- ALS MF, LMF, MP (총 3개)
- 총 3개의 추천을 진행하며 MF와 LMF 의 경우 콜드스타트 유저(신규 유저)인 경우 MP로 추천 진행

In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm

import scipy.sparse as sparse
import random
import implicit
from implicit.als import AlternatingLeastSquares as ALS

%cd /home/user_3/medistream-recsys/Script
from preprocessing import drop_columns,dict_to_column,dict_to_set,set_to_column,key_to_element

pd.set_option('display.max_rows', 300)
pd.set_option('display.max_columns', 100)

/home/user_3/medistream-recsys/Script


# 1.Dataload

In [44]:
# products name 확인 용
products_df = pd.read_json("/fastcampus-data/products/products.json")
products_df = key_to_element(['_id'],products_df)

100%|██████████| 5141/5141 [00:00<00:00, 449583.35it/s]


In [45]:
df = pd.read_json('/fastcampus-data/select_column_version_4.json')

In [46]:
date_state = "2022-05-12"

# paid orders만 가져오기
df['date_paid'] = pd.to_datetime(df['date_paid'])
df_only_paid = df[~df['date_paid'].isna()]
# 3개월치 데이터만 가져오기
df_date = df_only_paid[df_only_paid['date_paid'] > date_state]
# 취소 안된 것만 가져오기
complete_df = df_date[(df_date['paid'] == True) & (df_date['cancelled']==False)]
# 도서 카테고리만 가져오기
only_book = complete_df[complete_df['name'] == '도서']

# 유저가 중복으로 아이템 구매 삭제
df_duplicated_book = only_book.drop_duplicates(subset=['customer_id','product_ids'])
df_book = df_duplicated_book.sort_values(by='date_paid').reset_index(drop=True)

# 도서, 소모품 카테고리
# df_book = complete_df[complete_df['name'].isin(['도서','소모품'])].sort_values(by='date_paid')

In [47]:
# none 값 확인하기
df_book.isna().sum()

_id                0
date_created       0
regular_price      0
sale_price         0
three_months       0
date_paid          0
customer_id        0
paid               0
name_x             0
category_id_y      0
product_ids        0
quantity           0
price              0
price_total        0
age_group        905
한의사 여부             1
사업자 여부             1
cancelled          0
name               0
slug               0
dtype: int64

## 전체 데이터 EDA

In [48]:
print('중복 제거 전:',len(only_book), '중복 제거 후:',len(df_book))

중복 제거 전: 5931 중복 제거 후: 5910


In [49]:
print('전체 데이터 수:',len(df_book))

전체 데이터 수: 5910


In [50]:
print('아이템 수:',len(df_book.product_ids.unique()),'유저 수:',len(df_book.customer_id.unique()))

아이템 수: 252 유저 수: 2663


# 2.train test split
- 마지막 3주 분량을 test로 선정합니다.
- train 없는 test 아이템을 삭제 진행합니다.

In [51]:
from datetime import datetime, timedelta

# 마지막 날
df_book['date_paid'].max()

Timestamp('2022-09-13 08:51:40+0000', tz='UTC')

In [52]:
# 마지막 날 기준으로 test 전처리
df_book['date_paid'].max()-timedelta(weeks=3)

Timestamp('2022-08-23 08:51:40+0000', tz='UTC')

In [53]:
# train test 분리
date = '2022-08-23 08:51:40+0000'
train_before_preprocess = df_book[df_book['date_paid'] < date]
test_before_preprocess = df_book[df_book['date_paid'] >= date]

In [54]:
# # test 전처리 날짜 확인
# test_before_preprocess['date_paid'].max()-timedelta(weeks=1)

In [55]:
# test_del_date = '2022-09-06 08:51:40+0000'
# test_before_preprocess = test_before_preprocess[test_before_preprocess['date_paid']<=test_del_date]

In [56]:
# test 날짜 확인
print('첫날 :',test_before_preprocess['date_paid'].min(),'마지막 날:',test_before_preprocess['date_paid'].max())
print('test 총 기간:',test_before_preprocess['date_paid'].max()-test_before_preprocess['date_paid'].min())

첫날 : 2022-08-23 08:52:17+00:00 마지막 날: 2022-09-13 08:51:40+00:00
test 총 기간: 20 days 23:59:23


## 전체 아이템 중복 확인

In [57]:
# product_ids, name_x 수는 일치
len(df_book.product_ids.unique()), len(df_book.name_x.unique())

(252, 252)

In [58]:
# 중복 제거 후 수 비교 확인
# 252로 일치하여 문제 없음
len(df_book.drop_duplicates(subset=['product_ids','name_x']).name_x.unique())

252

## train test 아이템 중복 확인

In [59]:
len(train_before_preprocess.product_ids.unique()),len(test_before_preprocess.product_ids.unique())

(243, 130)

In [60]:
len(set(train_before_preprocess.product_ids.unique())-set(test_before_preprocess.product_ids.unique()))

122

In [61]:
# test 아이템에 train 없는 아이템 확인
len(set(test_before_preprocess.product_ids.unique())-set(train_before_preprocess.product_ids.unique()))

9

In [62]:
# test 만 있는 item 제거
only_test_items = set(test_before_preprocess.product_ids.unique())-set(train_before_preprocess.product_ids.unique())
if_prepro_test = test_before_preprocess[~test_before_preprocess['product_ids'].isin(only_test_items)]

In [63]:
test = test_before_preprocess.copy()

In [64]:
# train 변수 명 변경
train = train_before_preprocess.copy()

In [65]:
# test 전처리 진행했을 경우
print('원본 test 수:', len(test))
print('전처리 진행했을 경우 test 수:', len(if_prepro_test))

원본 test 수: 886
전처리 진행했을 경우 test 수: 651


# train test eda

### 전처리 전후 비교

In [66]:
print('train 전처리 전:',len(train_before_preprocess), 'train 전처리 후:',len(train))

train 전처리 전: 5024 train 전처리 후: 5024


In [67]:
print('test 전처리 전:',len(test_before_preprocess), 'test 전처리 후:',len(test))

test 전처리 전: 886 test 전처리 후: 886


### user 수 비교 

In [68]:
print('train 유저 수:',len(train.customer_id.unique()))

train 유저 수: 2323


In [69]:
print('test 유저 수:',len(test.customer_id.unique()))

test 유저 수: 672


In [70]:
# 신규 유저는 MP 같은 다른 방법으로 추천 진행해야 함
print('test 만 있는 신규 유저 :',len(set(test['customer_id'].unique())- set(train['customer_id'].unique())))

test 만 있는 신규 유저 : 340


### item 개수 비교

In [71]:
print('train 아이템 수 :',len(set(train.product_ids)), 'test 아이템 수 :',len(set(test.product_ids)))

train 아이템 수 : 243 test 아이템 수 : 130


In [72]:
print('train 만 있는 아이템 수:',  len(set(train.product_ids)-set(test.product_ids)))

train 만 있는 아이템 수: 122


In [73]:
print('test 만 있는 아이템 수:', len(set(test.product_ids) - set(train.product_ids)))

test 만 있는 아이템 수: 9


# 3. sparse matrix 만들기

## ALS MF Matrix

In [74]:
PdIds = train.product_ids.unique()

PdIdToIndex = {}
indexToPdId = {}

colIdx = 0

for PdId in PdIds:
    PdIdToIndex[PdId] = colIdx
    indexToPdId[colIdx] = PdId
    colIdx += 1
    
userIds = train.customer_id.unique()

userIdToIndex = {}
indexToUserId = {}

rowIdx = 0

for userId in userIds:
    userIdToIndex[userId] = rowIdx
    indexToUserId[rowIdx] = userId
    rowIdx += 1

import scipy.sparse as sp

rows = []
cols = []
vals = []

for row in train.itertuples():
    rows.append(userIdToIndex[row.customer_id])
    cols.append(PdIdToIndex[row.product_ids])
    vals.append(1)

purchase_sparse = sp.csr_matrix((vals, (rows, cols)), shape=(rowIdx,colIdx))

matrix = purchase_sparse.todense()
matrix

matrix([[1, 0, 0, ..., 0, 0, 0],
        [0, 1, 0, ..., 0, 0, 0],
        [0, 1, 0, ..., 0, 0, 0],
        ...,
        [0, 0, 0, ..., 0, 1, 0],
        [0, 0, 0, ..., 0, 1, 0],
        [0, 0, 0, ..., 0, 1, 0]])

### Most_popular_matrix

In [75]:
most_popular = train.groupby(['product_ids','name_x']).count()['customer_id'].reset_index()
most_popular

product_ids  \
0    5d13115e32026c0b35383897   
1    5d5373e94e77525ec5ca1135   
2    5d59ee854e77525ec5ca1212   
3    5d77b31f19efa30eb29143c9   
4    5d77b73619efa30eb29143ca   
5    5d78491b19efa30eb29143cc   
6    5d784bac19efa30eb29143d0   
7    5d784e1719efa30eb29143d1   
8    5d78505019efa30eb29143d2   
9    5d7852c519efa30eb29143d3   
10   5d930c1a0dabe405b156d610   
11   5d96811f0dabe405b156d83a   
12   5d9edd880dabe405b156e20f   
13   5d9edfe70dabe405b156e213   
14   5d9ff4170dabe405b156e45c   
15   5db8d6140dabe405b156f856   
16   5dc388aa2bb59605ca3cf794   
17   5dca53e32bb59605ca3cfc6b   
18   5dd397592bb59605ca3d01ea   
19   5ddc689c2bb59605ca3d0872   
20   5ddf84182bb59605ca3d0afe   
21   5de4c8804267e105dffeea5a   
22   5de8933a4267e105dffeeca9   
23   5dedb5e04267e105dffeef43   
24   5df720784267e105dffef934   
25   5e0032004267e105dffefd59   
26   5e1ec0fe4267e105dfff0942   
27   5e31324b4267e105dfff0e37   
28   5e3cb2904267e105dfff13b3   
29   5e3cfd1e4267e105dfff13d9   
30   5e3d00964267e105dfff13db   
31   5e3d01994267e105dfff13dc   
32   5e3d02de4267e105dfff13dd   
33   5e3d083c4267e105dfff13e3   
34   5e3d0a2f4267e105dfff13e6   
35   5e3d0da94267e105dfff13ed   
36   5e3d0f614267e105dfff13ef   
37   5e3d189d4267e105dfff13f2   
38   5e3d1d8d4267e105dfff13f3   
39   5e40b4e54267e105dfff1544   
40   5e40c62d4267e105dfff1559   
41   5e40cab84267e105dfff155b   
42   5e40cbc04267e105dfff155c   
43   5e40cc8b4267e105dfff155d   
44   5e40d6ea4267e105dfff1564   
45   5e40d9f24267e105dfff1565   
46   5e40dab74267e105dfff1566   
47   5e40dba04267e105dfff156a   
48   5e40ddc64267e105dfff1570   
49   5e42118d4267e105dfff17b0   
50   5e42133a4267e105dfff17b1   
51   5e4214414267e105dfff17b4   
52   5e4214f94267e105dfff17b8   
53   5e4215f84267e105dfff17b9   
54   5e4216cd4267e105dfff17bc   
55   5e423fa84267e105dfff181d   
56   5e42406f4267e105dfff1824   
57   5e4241204267e105dfff182a   
58   5e42421d4267e105dfff182b   
59   5e4242be4267e105dfff182c   
60   5e4244194267e105dfff1831   
61   5e4244c74267e105dfff1832   
62   5e425db84267e105dfff1855   
63   5e425e674267e105dfff1856   
64   5e425ef64267e105dfff1857   
65   5e4b427d4267e105dfff1de7   
66   5e4b43fa4267e105dfff1de8   
67   5e4b5a974267e105dfff1df8   
68   5e4b67d04267e105dfff1df9   
69   5e4b694b4267e105dfff1dfc   
70   5e4b69ed4267e105dfff1dfd   
71   5e4b6a984267e105dfff1dfe   
72   5e4b6b724267e105dfff1dff   
73   5e4b6c524267e105dfff1e04   
74   5e4b71e04267e105dfff1e12   
75   5e4b72fd4267e105dfff1e13   
76   5e4b73b14267e105dfff1e14   
77   5e4b76724267e105dfff1e1c   
78   5e4b84684267e105dfff1e26   
79   5e4ca4c14267e105dfff1e86   
80   5e4cadd84267e105dfff1e8a   
81   5e4cb11d4267e105dfff1e8d   
82   5e4cca6b4267e105dfff1e95   
83   5e4ccce44267e105dfff1e99   
84   5e4cd7554267e105dfff1ea2   
85   5e4cdc7e4267e105dfff1eab   
86   5e8adeeebfe4260944784c9d   
87   5e8ebf1abfe42609447850c6   
88   5e93e5d6bfe426094478533e   
89   5e9d478fbfe4260944785e62   
90   5ea28a26f1c8cc684388d214   
91   5ea643ec1a3ca36b96784a07   
92   5ecb7356e510ee503167b897   
93   5ecb7358e510ee503167b898   
94   5ef053ece510ee503167d19b   
95   5f06a1f3221c8b5f77c8de38   
96   5f06a1f4221c8b5f77c8de39   
97   5f06a1f5221c8b5f77c8de3a   
98   5f06a1f6221c8b5f77c8de3b   
99   5f06a1f7221c8b5f77c8de3d   
100  5f06a1fa221c8b5f77c8de40   
101  5f0bc3a9221c8b5f77c8df1c   
102  5f1550bf221c8b5f77c8f35b   
103  5f1550c2221c8b5f77c8f35d   
104  5f167b1e221c8b5f77c8f683   
105  5f167b1f221c8b5f77c8f685   
106  5f167b21221c8b5f77c8f686   
107  5f167b22221c8b5f77c8f687   
108  5f167b24221c8b5f77c8f689   
109  5f167b27221c8b5f77c8f690   
110  5f167b28221c8b5f77c8f692   
111  5f167b28221c8b5f77c8f693   
112  5f167b2a221c8b5f77c8f697   
113  5f167b2b221c8b5f77c8f698   
114  5f167b2d221c8b5f77c8f69d   
115  5f167b2e221c8b5f77c8f6a2   
116  5f167b2f221c8b5f77c8f6a4   
117  5f167b31221c8b5f77c8f6a9   
118  5f167b32221c8b5f77c8f6ac   
119  5f1e78fa221c8b5f77c9049e   
120  5f27d31c70973e5c1622

### Medistream_prediction_matrix
- 메디스트림 메디마켓에서 제공하는 정렬 추천 성능 비교를 위한 df 구현
- 인기도순, 최신순, 과거순, 높은 가격순, 낮은 가격순, 이름순 (총 6 가지)
- 각각 구현해보고 학습 모델 대비 성능 비교

In [76]:
medistream_prediction_df = train[['date_created','regular_price','sale_price','three_months','product_ids','name_x']]
medistream_prediction_preprop_df = medistream_prediction_df.drop_duplicates(subset=['product_ids'], ignore_index=True)
medistream_prediction_preprop_df['date_created'] = pd.to_datetime(medistream_prediction_preprop_df['date_created'])
# sale_prices가 0이면 regular_price 값으로 채워넣어야하는데 0이 없음(전처리 필요 무)

<ipython-input-76-4d4e36d189e6>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  medistream_prediction_preprop_df['date_created'] = pd.to_datetime(medistream_prediction_preprop_df['date_created'])


# Sparsity 확인

In [77]:
# Sparsity: 얼마나 비어있나?
matrix_size = purchase_sparse.shape[0]* purchase_sparse.shape[1]
num_purchases = len(purchase_sparse.nonzero()[0])
sparsity = 100 * (1 - (num_purchases / matrix_size))
sparsity

99.10999151444935

# 4. Model

## Model 학습 진행
- real test 만들기
- implict 라이브러리 사용(MF,LMF)
- MF 구현 모델 사용

In [78]:
# real test 
ground_trues = []
for user_id in test['customer_id'].unique():
    ground_trues.append({'id': user_id,\
    'items':list(test[test['customer_id']==user_id].product_ids)
    })

## ALS fit

In [79]:
als_model = ALS(factors=20, regularization=0.01, iterations = 50, random_state=42)
als_model.fit(purchase_sparse)

  0%|          | 0/50 [00:00<?, ?it/s]

In [80]:
# item, user vector 추출
als_item_factors = als_model.item_factors
als_user_factors = als_model.user_factors

In [81]:
# 각 shape 확인
als_item_factors.shape, als_user_factors.shape

((243, 20), (2323, 20))

## LMF fit

In [82]:
from implicit.lmf import LogisticMatrixFactorization as LMF

In [83]:
lmf_model = LMF(factors=20, regularization=0.001, iterations = 20, random_state=42)
lmf_model.fit(purchase_sparse)

  0%|          | 0/20 [00:00<?, ?it/s]

In [84]:
lmf_item_factors = lmf_model.item_factors
lmf_user_factors = lmf_model.user_factors

In [85]:
lmf_user_factors.shape, lmf_item_factors.shape

((2323, 22), (243, 22))

# 5. prediction

# ALS mf prediction

In [86]:
# 신규 유저인 경우 mp로 넣기
# 전체 도서에 대한 판매 만큼 정렬 후 넣기
most_popular_list = most_popular.sort_values(by='customer_id',ascending=False).index

# test 예측값, 이미 구매 했을 경우 제외
als_predict_list = []
for user_id in test['customer_id'].unique():
    try:
        result = als_model.recommend(userIdToIndex[user_id], purchase_sparse[userIdToIndex[user_id]], N=100)
        als_predict_list.append({'id':user_id ,'items':[indexToPdId[num] for num in result[0]]})
    except:
        train_purchase_list = list(train[train['customer_id']==user_id].product_ids)
        als_predict_list.append({'id':user_id ,'items':[most_popular.product_ids.loc[num] for num in most_popular_list \
                                                            if most_popular.product_ids.loc[num] not in train_purchase_list \
                                                            ]})
        
# 100 개만 예측하기
for idx, pred_list in enumerate(als_predict_list):
    als_predict_list[idx]['items'] = pred_list['items'][:100]

# LMF prediction

In [87]:
# 신규 유저 mp로 넣기
most_popular_list = most_popular.sort_values(by='customer_id',ascending=False).index

# test 예측값
lmf_predict_list = []
for user_id in test['customer_id'].unique():
    try:
        result = lmf_model.recommend(userIdToIndex[user_id], purchase_sparse[userIdToIndex[user_id]], N=100)
        lmf_predict_list.append({'id':user_id ,'items':[indexToPdId[num] for num in result[0]]})
    except:
        train_purchase_list = list(train[train['customer_id']==user_id].product_ids)
        lmf_predict_list.append({'id':user_id ,'items':[most_popular.product_ids.loc[num] for num in most_popular_list \
                                                            if most_popular.product_ids.loc[num] not in train_purchase_list \
                                                            ]})
        
# 100 개만 예측하기
for idx, pred_list in enumerate(lmf_predict_list):
    lmf_predict_list[idx]['items'] = pred_list['items'][:100]

# most popular prediction

In [88]:
# 전체 도서에 대한 판매 만큼 정렬 후 넣기
most_popular_list = most_popular.sort_values(by='customer_id',ascending=False).index

# test 예측값, 이미 구매 했을 경우 제외
predict_popular_list = []
for user_id in test['customer_id'].unique():
    train_purchase_list = list(train[train['customer_id']==user_id].product_ids)
    predict_popular_list.append({'id':user_id ,'items':[most_popular.product_ids.loc[num] for num in most_popular_list \
                                                            if most_popular.product_ids.loc[num] not in train_purchase_list \
                                                            ]})

# 100 개만 예측하기
for idx, pred_list in enumerate(predict_popular_list):
    predict_popular_list[idx]['items'] = pred_list['items'][:100]

# medistream prediction
- 메디스트림 메디마켓에서 제공하는 정렬 추천 성능 비교
- 인기도순, 최신순, 과거순, 높은 가격순, 낮은 가격순, 이름순 (총 6 가지)
- 각각 구현해보고 학습 모델 대비 성능 비교

In [89]:
# 인기도순
medistream_popular_list = medistream_prediction_preprop_df.sort_values(by='three_months', ascending=False).index
# 최신순
medistream_latest_list = medistream_prediction_preprop_df.sort_values(by='date_created', ascending=False).index
# 오랜된 순
medistream_oldest_list = medistream_prediction_preprop_df.sort_values(by='date_created', ascending=True).index
# 높은 가격 순
medistream_high_price_list = medistream_prediction_preprop_df.sort_values(by='sale_price', ascending=False).index
# 낮은 가격 순
medistream_low_price_list = medistream_prediction_preprop_df.sort_values(by='sale_price', ascending=True).index
# 이름 순
medistream_name_sort_list = medistream_prediction_preprop_df.sort_values(by='name_x',ascending=True).index

def medistream_prediction_method(predict_num:int ,medi_predict_list:list)->list:
    medistream_predict_list = []
    for user_id in test['customer_id'].unique():
        medistream_predict_list.append({'id':user_id ,'items':[medistream_prediction_preprop_df.product_ids.loc[num] \
                                                                       for num in medi_predict_list]})

    # 100 개만 예측하기
    for idx, pred_list in enumerate(medistream_predict_list):
        medistream_predict_list[idx]['items'] = pred_list['items'][:predict_num]
        
    return medistream_predict_list

In [90]:
medistream_predict_popular_list = medistream_prediction_method(100, medistream_popular_list)
medistream_predict_latest_list = medistream_prediction_method(100, medistream_latest_list)
medistream_predict_oldest_list = medistream_prediction_method(100, medistream_oldest_list)
medistream_predict_high_price_list = medistream_prediction_method(100, medistream_high_price_list)
medistream_predict_low_price_list = medistream_prediction_method(100, medistream_low_price_list)
medistream_predict_name_sort_list = medistream_prediction_method(100, medistream_name_sort_list)

# 6. evaluation

## NDCG 평가지표

In [91]:
class CustomEvaluator:
    # relavence 모두 1로 동일하게 봄
    def _idcg(self, l):
        return sum((1.0 / np.log(i + 2) for i in range(l)))
    

    def __init__(self):
        self._idcgs = [self._idcg(i) for i in range(1000)]
    '''
    idcgs 예시, item 3개 추천되므로 3.074281787960283 가 됩니다.
    [0, 1.4426950408889634, 2.352934267515801, 3.074281787960283]
    '''

    def _ndcg(self, gt, rec):
        dcg = 0.0
        for i, r in enumerate(rec):
            if r in gt:
                dcg += 1.0 / np.log(i + 2)

        return dcg / self._idcgs[len(gt)]

    def _eval(self, gt_list, rec_list):
        gt_dict = {g["id"]: g for g in gt_list}
        ndcg_score = 0.0

        for rec in rec_list:
            gt = gt_dict[rec["id"]]
            ndcg_score += self._ndcg(gt["items"], rec["items"])


        ndcg_score = ndcg_score / len(rec_list)


        return ndcg_score

    def evaluate(self, gt_list, rec_list):
        try:
            ndcg_score = self._eval(gt_list, rec_list)
            print(f"nDCG: {ndcg_score:.6}")
        except Exception as e:
            print(e)


# ALS NDCG

In [92]:
# ALS 
evaluator = CustomEvaluator()
evaluator.evaluate(ground_trues, als_predict_list)

nDCG: 0.139074


In [93]:
len(als_predict_list),len(ground_trues)

(672, 672)

In [94]:
# 아이템 맞춘 개수
cnt = 0
for gt, pred_list in zip(ground_trues, als_predict_list):
    for pred in pred_list['items']:
        if pred in gt['items']:
            cnt += 1
cnt

442

# LMF NDCG

In [95]:
# ALS 
evaluator = CustomEvaluator()
evaluator.evaluate(ground_trues, lmf_predict_list)

nDCG: 0.184686


In [96]:
len(lmf_predict_list),len(ground_trues)

(672, 672)

In [97]:
# 아이템 맞춘 개수
cnt = 0
for gt, pred_list in zip(ground_trues, lmf_predict_list):
    for pred in pred_list['items']:
        if pred in gt['items']:
            cnt += 1
cnt

527

# most popular NDCG

In [98]:
# most popular
evaluator = CustomEvaluator()
evaluator.evaluate(ground_trues, predict_popular_list)

nDCG: 0.184859


In [99]:
len(predict_popular_list),len(ground_trues)

(672, 672)

In [100]:
# 아이템 맞춘 개수
cnt = 0
for gt, pred_list in zip(ground_trues, predict_popular_list):
    for pred in pred_list['items']:
        if pred in gt['items']:
            cnt += 1
cnt

546

## medistream prediction NDCG

In [101]:
def medistream_prediction(ground_trues:list, predict_list:list):
    evaluator = CustomEvaluator()
    ndcg = evaluator._eval(ground_trues, predict_list)
    
    assert len(predict_list) == len(ground_trues)
    
    cnt = 0
    for gt, pred_list in zip(ground_trues, predict_list):
        for pred in pred_list['items']:
            if pred in gt['items']:
                cnt += 1
    return ndcg, cnt

In [102]:
medistream_predict_score = {'medistream_predict':['medi_popular','latest','oldest','higt_price','low_price','name_sort'], \
                            'ndcg':[], 'cnt':[]}

medistream_predict_list = [medistream_predict_popular_list, medistream_predict_latest_list, medistream_predict_oldest_list,\
                          medistream_predict_high_price_list, medistream_predict_low_price_list, medistream_predict_name_sort_list]

for medistream_predict in medistream_predict_list:
    ndcg, cnt = medistream_prediction(ground_trues, medistream_predict)
    medistream_predict_score['ndcg'].append(ndcg)
    medistream_predict_score['cnt'].append(cnt)
pd.DataFrame(medistream_predict_score)    

medistream_predict      ndcg  cnt
0       medi_popular  0.280721  600
1             latest  0.406678  452
2             oldest  0.041424  159
3         higt_price  0.026910  138
4          low_price  0.087135  408
5          name_sort  0.033603  159

# 7. hyper parameter tuning

## 7-1. ALS MF hypter parameter tuning

In [103]:
als_mf_hyper_parameter = {'factor':[],'regularization':[],'iteration':[],'NDCG':[]}

factors = [5,10,15,20]
regularizations = [0.01,0.005]
iterations = [5,10,15,20,25,30,40,50]

for factor in factors:
    for regularization in regularizations:
        for iteration in iterations:
            als_model = ALS(factors=factor, regularization=regularization, iterations = iteration, random_state=42)
            als_model.fit(purchase_sparse, show_progress=False)
            
            # 신규 유저인 경우 mp로 넣기
            # 전체 도서에 대한 판매 만큼 정렬 후 넣기
            most_popular_list = most_popular.sort_values(by='customer_id',ascending=False).index

            # test 예측값, 이미 구매 했을 경우 제외
            als_predict_list = []
            for user_id in test['customer_id'].unique():
                try:
                    result = als_model.recommend(userIdToIndex[user_id], purchase_sparse[userIdToIndex[user_id]], N=100)
                    als_predict_list.append({'id':user_id ,'items':[indexToPdId[num] for num in result[0]]})
                except:
                    train_purchase_list = list(train[train['customer_id']==user_id].product_ids)
                    als_predict_list.append({'id':user_id ,'items':[most_popular.product_ids.loc[num] for num in most_popular_list \
                                                                        if most_popular.product_ids.loc[num] not in train_purchase_list \
                                                                        ]})

            # 100 개만 예측하기
            for idx, pred_list in enumerate(als_predict_list):
                als_predict_list[idx]['items'] = pred_list['items'][:100]
                
            # ALS 
            evaluator = CustomEvaluator()
            ndcg = evaluator._eval(ground_trues, als_predict_list)
            
            als_mf_hyper_parameter['factor'].append(factor)
            als_mf_hyper_parameter['regularization'].append(regularization)
            als_mf_hyper_parameter['iteration'].append(iteration)
            als_mf_hyper_parameter['NDCG'].append(ndcg)

In [104]:
pd.DataFrame(als_mf_hyper_parameter).sort_values(by='NDCG',ascending=False).head()

factor  regularization  iteration      NDCG
26      10           0.005         15  0.200100
27      10           0.005         20  0.199544
28      10           0.005         25  0.197715
22      10           0.010         40  0.197100
31      10           0.005         50  0.196771

## 7-2. LMF hypter parameter tuning

In [105]:
lmf_hyper_parameter = {'factor':[],'regularization':[],'iteration':[],'NDCG':[]}

factors = [5,10,15,20]
regularizations = [0.01,0.005]
iterations = [5,10,15,20,25,30,40,50]

for factor in factors:
    for regularization in regularizations:
        for iteration in iterations:
            lmf_model = LMF(factors=factor, regularization=regularization, iterations = iteration, random_state=42)
            lmf_model.fit(purchase_sparse, show_progress=False)
            
            # 신규 유저 mp로 넣기
            most_popular_list = most_popular.sort_values(by='customer_id',ascending=False).index

            # test 예측값
            lmf_predict_list = []
            for user_id in test['customer_id'].unique():
                try:
                    result = lmf_model.recommend(userIdToIndex[user_id], purchase_sparse[userIdToIndex[user_id]], N=100)
                    lmf_predict_list.append({'id':user_id ,'items':[indexToPdId[num] for num in result[0]]})
                except:
                    train_purchase_list = list(train[train['customer_id']==user_id].product_ids)
                    lmf_predict_list.append({'id':user_id ,'items':[most_popular.product_ids.loc[num] for num in most_popular_list \
                                                                        if most_popular.product_ids.loc[num] not in train_purchase_list \
                                                                        ]})

            # 100 개만 예측하기
            for idx, pred_list in enumerate(lmf_predict_list):
                lmf_predict_list[idx]['items'] = pred_list['items'][:100]
                
            # LMF
            evaluator = CustomEvaluator()
            ndcg = evaluator._eval(ground_trues, lmf_predict_list)
            
            lmf_hyper_parameter['factor'].append(factor)
            lmf_hyper_parameter['regularization'].append(regularization)
            lmf_hyper_parameter['iteration'].append(iteration)
            lmf_hyper_parameter['NDCG'].append(ndcg)

In [106]:
pd.DataFrame(lmf_hyper_parameter).sort_values(by='NDCG',ascending=False).head()

factor  regularization  iteration      NDCG
44      15           0.005         25  0.206777
45      15           0.005         30  0.205400
36      15           0.010         25  0.203824
37      15           0.010         30  0.202828
47      15           0.005         50  0.201626

# 8. 결론

- als mf : 0.200100 (factor: 10, regularization: 0.005, iteration: 15)
- lmf : 0.206777    (factor: 15, regularization: 0.005, iteration: 25)
- mp : 0.184859     

- test 신규 아이템을 포함하여 예측을 진행했다고 하더라도 나쁘지 않은 점수가 나옴

# 9. 추천된 items 확인

In [107]:
products_df = pd.read_json("/fastcampus-data/products/products.json")
products_df = key_to_element(['_id'],products_df)

100%|██████████| 5141/5141 [00:00<00:00, 698960.03it/s]


In [108]:
# pred_item, rea_item 비교
def pred_real_dataframe(user_num):
    pred_items_names = []
    predict_dict = als_predict_list[user_num]['items']
    for item in predict_dict:
        pred_items_names.append(products_df[products_df['_id'] == item].meta_title.unique())

    real_items_names = []
    trues_dict = ground_trues[user_num]['items']
    for item in trues_dict:
        real_items_names.append(products_df[products_df['_id'] == item].meta_title.unique())
    return pd.DataFrame({'pred_item':pred_items_names,'real_item':real_items_names})
    

In [109]:
# train_item, pred_item, real_item 비교
def train_pred_items(user_nums):
    train_pred_items_df = pd.DataFrame(columns=['train_item','pred_item'])
    for user_num in range(1,user_nums):
        train_item_names = []
        for idx in grouped_purchased[grouped_purchased['customer_id']==ground_trues[user_num]['id']].product_ids:
            train_item_names.append(products_df[products_df['_id'] == idx].meta_title.unique()[0])

        pred_items_names = []
        predict_dict = als_predict_list[user_num]['items']
        for item in predict_dict:
            pred_items_names.append(products_df[products_df['_id'] == item].meta_title.unique())


        
        train_pred_items_df.loc[user_num,'train_item'] = train_item_names
        train_pred_items_df.loc[user_num,'pred_item'] = pred_items_names
    return train_pred_items_df


In [110]:
train_pred_items(13)

NameError: name 'grouped_purchased' is not defined

In [ ]:
# 예측 유저 구매 횟수 확인
pd.DataFrame(purchase_sparse[1].todense()).T.value_counts()

In [111]:
print('첫날 :',train['date_paid'].min(),'마지막 날:',train['date_paid'].max())
print('train 총 기간:',train['date_paid'].max()-train['date_paid'].min())
print('______________________________________________________')
print('첫날 :',test['date_paid'].min(),'마지막 날:',test['date_paid'].max())
print('test 총 기간:',test['date_paid'].max()-test['date_paid'].min())
print('______________________________________________________')
print('train 데이터수:', len(train))
print('______________________________________________________')
print('train 유저수:',len(set(train.customer_id)))
print('______________________________________________________')
print('test 데이터수:',len(test))
print('______________________________________________________')
print('test 유저수:',len(set(test.customer_id)))
print('______________________________________________________')
print('test 신규 유저 수:',len(set(test['customer_id'].unique())- set(train['customer_id'].unique())))
print('______________________________________________________')
print('test 신규 아이템 수:',len(set(test_before_preprocess.product_ids.unique())-set(train_before_preprocess.product_ids.unique())))
# test 전처리 진행했을 경우
print('______________________________________________________')
print('원본 test 수:', len(test))
print('______________________________________________________')
print('전처리 진행했을 경우 test 수:', len(if_prepro_test))
print('______________________________________________________')
print('mf')
display(pd.DataFrame(als_mf_hyper_parameter).sort_values(by='NDCG',ascending=False).head(2))
print('______________________________________________________')
print('lmf')
display(pd.DataFrame(lmf_hyper_parameter).sort_values(by='NDCG',ascending=False).head(2))
print('______________________________________________________')
print('mp')
evaluator.evaluate(ground_trues, predict_popular_list)
print('______________________________________________________')
display(pd.DataFrame(medistream_predict_score))

첫날 : 2022-05-12 00:10:00.012000+00:00 마지막 날: 2022-08-23 08:51:14+00:00
train 총 기간: 103 days 08:41:13.988000
______________________________________________________
첫날 : 2022-08-23 08:52:17+00:00 마지막 날: 2022-09-13 08:51:40+00:00
test 총 기간: 20 days 23:59:23
______________________________________________________
train 데이터수: 5024
______________________________________________________
train 유저수: 2323
______________________________________________________
test 데이터수: 886
______________________________________________________
test 유저수: 672
______________________________________________________
test 신규 유저 수: 340
______________________________________________________
test 신규 아이템 수: 9
______________________________________________________
원본 test 수: 886
______________________________________________________
전처리 진행했을 경우 test 수: 651
______________________________________________________
mf


factor  regularization  iteration      NDCG
26      10           0.005         15  0.200100
27      10           0.005         20  0.199544

______________________________________________________
lmf


factor  regularization  iteration      NDCG
44      15           0.005         25  0.206777
45      15           0.005         30  0.205400

______________________________________________________
mp
nDCG: 0.184859
______________________________________________________


medistream_predict      ndcg  cnt
0       medi_popular  0.280721  600
1             latest  0.406678  452
2             oldest  0.041424  159
3         higt_price  0.026910  138
4          low_price  0.087135  408
5          name_sort  0.033603  159